### ISSUE TITLE
Finetuning and Config used for HPS

### ISSUE THREAD

**Issue by anonymous-atom** [mihirp1998/AlignProp#17]
@mihirp1998 
I was tryin to finetune Stable Diffusion 1.5 using your HPS reward function and the hps.sh training script, I used batch size of 1 but still the training seems to get completed very quickly, 50 epochs just took 2-4 minutes.

```
def from_file(path, low=None, high=None):
    prompts = _load_lines(path)[low:high]
    return random.choice(prompts), {}
```

And here you are trying to just use batch_size number of prompts ? I am using batch_size of 2 on 1 A100 GPU to test the script.
```
    def _generate_samples(self, batch_size, with_grad=True, prompts=None):
        """
        Generate samples from the model

        Args:
            batch_size (int): Batch size to use for sampling
            with_grad (bool): Whether the generated RGBs should have gradients attached to it.

        Returns:
            prompt_image_pairs (Dict[Any])
        """
        prompt_image_pairs = {}

        sample_neg_prompt_embeds = self.neg_prompt_embed.repeat(batch_size, 1, 1)

        if prompts is None:
            prompts, prompt_metadata = zip(*[self.prompt_fn() for _ in range(batch_size)])
```

Your help will mean a lot!

**Comment by mihirp1998** [comment]
Can u check whether the training is actually getting run ? if not why is it skipping the training loop

**Comment by anonymous-atom** [comment]
It's working but I think it's only using few examples from the prompt file.

**Comment by mihirp1998** [comment]
i think random.choice is uniform over all prompts, i'm not sure what's the bug here. If you find it let me know.

**Comment by anonymous-atom** [comment]
Yeah Sure!

**Comment by anonymous-atom** [comment]
Also while I was tryin to train with a custom loss function, the models seem to collapse very early unless I adjust the learning rate. is this the expected behaviour ?

**Comment by anonymous-atom** [comment]
@mihirp1998 Sorry to tag you again, but can you let me know how much time it took per epoch on your 4 A100 GPU's ?

Clear Skies!

**Comment by mihirp1998** [comment]
On Aesthetics 2-3 minutes per epoch

**Comment by anonymous-atom** [comment]
I wanted to confirm are you using all 750 prompts in hps_v2_all.txt file for 1 single epoch ?

**Comment by anonymous-atom** [comment]
Hi @mihirp1998 , so this is where I am confused:

So you used step() function to do 1 training step/ 1 epoch ?

```python
    def train(self, epochs: Optional[int] = None):
        """
        Train the model for a given number of epochs
        """
        global_step = 0
        if epochs is None:
            epochs = self.config.num_epochs
        for epoch in range(self.first_epoch, epochs):
            global_step = self.step(epoch, global_step)
```


And here in step() function, it only seems to finetine on **_num_gpus * batch_size * train_gradient_accumulation_steps_** number of images, am I missing something ? What if someone used just 1 GPU to train ?

```python

    def step(self, epoch: int, global_step: int):
   
        info = defaultdict(list)
        print(f"Epoch: {epoch}, Global Step: {global_step}")

        self.sd_pipeline.unet.train()

        for _ in range(self.config.train_gradient_accumulation_steps):
            with self.accelerator.accumulate(self.sd_pipeline.unet), self.autocast(), torch.enable_grad():
                prompt_image_pairs = self._generate_samples(
                    batch_size=self.config.train_batch_size,
                )
                
```

**Comment by findalexli** [comment]
following this thread. this is critical setup otherwise other folks cannot reproduce this

**Comment by mihirp1998** [comment]
> Hi @mihirp1998 , so this is where I am confused:
> 
> So you used step() function to do 1 training step/ 1 epoch ?
> 
> ```python
>     def train(self, epochs: Optional[int] = None):
>         """
>         Train the model for a given number of epochs
>         """
>         global_step = 0
>         if epochs is None:
>             epochs = self.config.num_epochs
>         for epoch in range(self.first_epoch, epochs):
>             global_step = self.step(epoch, global_step)
> ```
> 
> And here in step() function, it only seems to finetine on **_num_gpus * batch_size * train_gradient_accumulation_steps_** number of images, am I missing something ? What if someone used just 1 GPU to train ?
> 
> ```python
>     def step(self, epoch: int, global_step: int):
>    
>         info = defaultdict(list)
>         print(f"Epoch: {epoch}, Global Step: {global_step}")
> 
>         self.sd_pipeline.unet.train()
> 
>         for _ in range(self.config.train_gradient_accumulation_steps):
>             with self.accelerator.accumulate(self.sd_pipeline.unet), self.autocast(), torch.enable_grad():
>                 prompt_image_pairs = self._generate_samples(
>                     batch_size=self.config.train_batch_size,
>                 )
>                 
> ```

Yes in this codebase step and epoch are equivalent, it's difficult to define an epoch as there is no dataset i'm training on.


If someone used one gpu for training and wants to maintain the batchsize i'm using, they should increase the accum_steps as i mention here:

https://github.com/mihirp1998/AlignProp/blob/5e950b3f16ded622df15f4bea2eec93f88962f2b/hps.sh#L1

**Comment by Anirudh257** [comment]
Hi @mihirp1998 I ran your code but was unable to reproduce the result for HPS. I got 0.295 but the reported results are 0.330. I followed the steps to setup the code given in the repo and ran 

> bash hps.sh

for `100 epochs`.

My graph looks like:
![image](https://github.com/user-attachments/assets/ec6b6f87-0740-4a7b-8b85-ed929cc5e6dc)

**Comment by mihirp1998** [comment]
![Screenshot 2024-12-01 at 5 00 33 PM](https://github.com/user-attachments/assets/b92e5a5a-dd2d-401e-9d05-89f027c8dde2)

here's one of of my old curves. Can u maybe try with the main branch? https://github.com/mihirp1998/AlignProp/tree/main

I actually forgot to reproduce hps on the trl_main branch

**Comment by Anirudh257** [comment]
@mihirp1998 Thanks, is https://github.com/mihirp1998/AlignProp/tree/main

 containing the correct code for reproducing Aesthetics score as well?
 
I am unable to reproduce that too. I got 8.41 (reported is 8.94)

My graphs:

![image](https://github.com/user-attachments/assets/87578c52-9b45-44fe-bbdc-9c887239f9f4)

**Comment by mihirp1998** [comment]
i think if u just train them for longer u can get a score of 10, however the images won't follow the text anymore.

For instance for aesthetic results i do not normally train over 10 epochs.

So i would not suggest trying to get high scores, but early stop when they look reasonably good

**Comment by Anirudh257** [comment]
Okay, I was just following the paper where you mention 100 epochs of training.

**Comment by mihirp1998** [comment]
I think there might be some variance via seed etc, but they should be in similar range. But again 100 epochs was just to report the results in the table and is not necessarily the epoch that gives the best visual results.

**Comment by Anirudh257** [comment]
Thanks, that's true.
